File: packages/client/py/spellguard_client/dependencies.py
# SPDX-License-Identifier: Apache-2.0

"""
spellguard_client.dependencies — agent-side helpers for reporting
lockfile / dependency snapshots to Management's advisory pipeline.

Mirrors :mod:`spellguard_client/dependencies.ts`.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from typing import Any, List, Literal, Optional

import httpx

SUPPORTED_LOCKFILES: tuple[str, ...] = (
    "pnpm-lock.yaml",
    "pnpm-lock.yml",
    "yarn.lock",
    "package-lock.json",
    "requirements.txt",
    "poetry.lock",
    "Cargo.lock",
    "go.sum",
    "sbom.cdx.json",
    "cyclonedx.json",
    "sbom.json",
)


@dataclass
class LockfileFile:
    filename: str
    content: str


@dataclass
class ParsedDependency:
    ecosystem: str
    package_name: str
    package_version: str
    dep_type: Literal["runtime", "dev", "transitive"]


def read_lockfile_from_dir(directory: str) -> Optional[LockfileFile]:
    """Locate and read the first supported lockfile in *directory*.

    Returns ``None`` if no lockfile is present; the caller decides
    whether to skip the upload or fail loudly.
    """
    for candidate in SUPPORTED_LOCKFILES:
        path = os.path.join(directory, candidate)
        if os.path.isfile(path):
            with open(path, "r", encoding="utf-8") as f:
                return LockfileFile(filename=candidate, content=f.read())
    return None


async def report_dependencies(
    *,
    management_url: str,
    agent_id: str,
    agent_token: str,
    lockfile: Optional[LockfileFile] = None,
    dependencies: Optional[List[ParsedDependency]] = None,
    lockfile_hash: Optional[str] = None,
    timeout_seconds: float = 30.0,
) -> dict[str, Any]:
    """POST the agent's lockfile / dependencies to Management.

    Pass either ``lockfile`` (parser-driven ingestion) or ``dependencies``
    + ``lockfile_hash`` (caller pre-parsed). Returns the server's parse
    summary; raises ``RuntimeError`` on non-2xx responses.
    """
    body: dict[str, Any]
    if lockfile is not None:
        body = {
            "lockfile": {
                "filename": lockfile.filename,
                "content": lockfile.content,
            }
        }
    elif dependencies is not None and lockfile_hash is not None:
        body = {
            "dependencies": [
                {
                    "ecosystem": d.ecosystem,
                    "packageName": d.package_name,
                    "packageVersion": d.package_version,
                    "depType": d.dep_type,
                }
                for d in dependencies
            ],
            "lockfileHash": lockfile_hash,
        }
    else:
        raise ValueError(
            "report_dependencies: pass either lockfile= or "
            "dependencies= + lockfile_hash="
        )
    url = f"{management_url.rstrip('/')}/v1/agents/{agent_id}/dependencies"
    async with httpx.AsyncClient(timeout=timeout_seconds) as client:
        response = await client.post(
            url,
            headers={
                "Authorization": f"Bearer {agent_token}",
                "Content-Type": "application/json",
            },
            json=body,
        )
    if response.status_code >= 400:
        raise RuntimeError(
            f"report_dependencies failed: {response.status_code} "
            f"{response.reason_phrase} — {response.text}"
        )
    return response.json()


__all__ = [
    "SUPPORTED_LOCKFILES",
    "LockfileFile",
    "ParsedDependency",
    "read_lockfile_from_dir",
    "report_dependencies",
]

File: packages/client/ts/src/dependencies.ts
// SPDX-License-Identifier: Apache-2.0
//
// Helper for agents to report their lockfile / dependency snapshot to
// Management's advisory pipeline. Designed to be called once on agent
// startup (or at deploy time via a CI script) so the supply-chain
// detection pipeline has up-to-date input.
//
// Two layers:
//   - `readLockfileFromDir(dir)` — Node.js-only convenience that locates
//     a lockfile in a directory and reads it. Workers callers can't use
//     this (no `fs` access); they should bundle the lockfile content at
//     build time and pass it directly to `reportDependencies`.
//   - `reportDependencies(opts)` — POSTs the lockfile content (or pre-
//     parsed dependencies) to `${managementUrl}/v1/agents/:agentId/dependencies`.
//
// Both are tree-shakeable; agents only pay for what they import.

export interface LockfileFile {
  filename: string;
  content: string;
}

/**
 * Lockfile filenames the management-side parser recognizes, ordered by
 * preference (project lockfiles first, then Python, then Rust/Go, then
 * SBOM fallback).
 */
export const SUPPORTED_LOCKFILES = [
  'pnpm-lock.yaml',
  'pnpm-lock.yml',
  'yarn.lock',
  'package-lock.json',
  'requirements.txt',
  'poetry.lock',
  'Cargo.lock',
  'go.sum',
  'sbom.cdx.json',
  'cyclonedx.json',
  'sbom.json',
] as const;

/**
 * Locate and read the first supported lockfile in `dir`. Walks the
 * `SUPPORTED_LOCKFILES` list in order and returns the first match.
 * Returns `null` when no lockfile is present (caller decides whether
 * to skip the upload or fail loudly).
 *
 * Node.js-only: imports `node:fs` lazily so the function tree-shakes
 * out of Workers bundles. Callers that target Workers should pass the
 * lockfile content via build-time bundling and call
 * `reportDependencies` directly.
 */
export async function readLockfileFromDir(
  dir: string,
): Promise<LockfileFile | null> {
  // Lazy import keeps this out of Workers bundles. The dynamic specifier
  // also avoids static analyzers that flag `node:` imports in Workers
  // builds (they tree-shake when the function isn't called).
  const fs = await import('node:fs');
  const path = await import('node:path');
  for (const candidate of SUPPORTED_LOCKFILES) {
    const fullPath = path.join(dir, candidate);
    if (fs.existsSync(fullPath)) {
      const content = fs.readFileSync(fullPath, 'utf-8');
      return { filename: candidate, content };
    }
  }
  return null;
}

export interface ReportDependenciesOptions {
  managementUrl: string;
  agentId: string;
  /**
   * The agent's bearer token — typically the management agent secret.
   * `requireAuthOrApiKey` on the route accepts either a user JWT or an
   * agent token, so this works for both manual (CI) and runtime calls.
   */
  agentToken: string;
  /**
   * Either a raw lockfile (for parser-driven ingestion) or pre-parsed
   * dependency entries with the source lockfile's hash.
   */
  lockfile?: LockfileFile;
  dependencies?: ParsedDependency[];
  lockfileHash?: string;
  /** Override fetch (mostly for tests). */
  fetchImpl?: typeof fetch;
}

export interface ParsedDependency {
  ecosystem: string;
  packageName: string;
  packageVersion: string;
  depType: 'runtime' | 'dev' | 'transitive';
}

export interface ReportDependenciesResult {
  format: string;
  upserted: number;
  newAlerts: number;
  lockfileHash: string;
}

/**
 * POST the agent's lockfile / dependencies to Management. Returns the
 * server's parse summary. Throws on non-2xx responses; caller decides
 * whether to log-and-continue or hard-fail.
 */
export async function reportDependencies(
  opts: ReportDependenciesOptions,
): Promise<ReportDependenciesResult> {
  const { managementUrl, agentId, agentToken, fetchImpl = fetch } = opts;
  let body: Record<string, unknown>;
  if (opts.lockfile) {
    body = { lockfile: opts.lockfile };
  } else if (opts.dependencies && opts.lockfileHash) {
    body = { dependencies: opts.dependencies, lockfileHash: opts.lockfileHash };
  } else {
    throw new Error(
      'reportDependencies: pass either {lockfile} or {dependencies, lockfileHash}',
    );
  }
  const url = `${managementUrl.replace(/\/$/, '')}/v1/agents/${encodeURIComponent(agentId)}/dependencies`;
  const response = await fetchImpl(url, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${agentToken}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  });
  if (!response.ok) {
    const detail = await response.text().catch(() => '');
    throw new Error(
      `reportDependencies failed: ${response.status} ${response.statusText}${detail ? ` — ${detail}` : ''}`,
    );
  }
  return (await response.json()) as ReportDependenciesResult;
}
